### ISSUE TITLE
Subnet with empty validator set can be activated

### ISSUE THREAD

**Issue by dnkolegov** [consensus-shipyard/ipc-solidity-actors#71]
Let's say `minActivationCollateral` = 10, validator 1 stakes 3, validator 2 stakes 7.

After validator 1 calling `join`: `totalStake` = 3, `validatorSet` = {}
After validator 2 calling `join`: `totalStake` = 10, `validatorSet` = {} but since `totalStake >= minActivationCollateral` then `addStake` is called and changes the status of the subnet to `Active`. Then subnet actor also changes the subnet status to `Active`.

As a result, the subnet with an empty validator set is active.

**Comment by dnkolegov** [comment]
At present, "Active" means that a subnet has enough collateral. In the current design, even a subnet without validators can be active. This is also implemented in Gateway's `register` function that activates the Subnet. Also, the Gateway actor knows nothing about the number of validators in a subnet.
If we want explicitly signal to Gateway when a subnet should be activated/deactivated, we should use calls like `activateSubnet` and `deactivateSubnet` of the Gateway actor.

Closing for now.

**Comment by dnkolegov** [comment]
Another problem with the current design is that the subnet collateral can fall behind the minimum and cause the subnet to be deactivated. However, if every single validator has to put up at least as much as the minimum required collateral, then falling behind the threshold means all validators have abandoned the subnet.

https://github.com/consensus-shipyard/ipc-solidity-actors/pull/201#discussion_r1325783577

**Comment by aakoshh** [comment]
> After validator 1 calling join: totalStake = 3, validatorSet = {}

That sounds like a bug, shouldn't `join` add the guy to the `validatorSet`?

>  In the current design, even a subnet without validators can be active. 

That sounds strange, who put the collateral down if there are no validators? How does this design make any sense?

**Comment by dnkolegov** [comment]
> According to this implementation, it should not since the validator's stake is less than `minActivationCollateral`.

I agree that this is a bug the problem is that we do not have a spec.

**Comment by adlrocha** [comment]
I have an alternative view of this issue (I am not claiming it is the right one, but is how I was thinking about it when I first implemented this). 
- The gateway is the gatekeeper of IPC. It is responsible for keeping the collateral for the subnet and enforcing a set of minimum requirements for all its child subnets. The rules are simple right now: if a child subnet has enough collateral according to the threshold set for this IPC parent, the subnet is `Active` and can interact with IPC, if not, it is in `Inactive`. 
- On top of this, developers deploy their own subnet actors to govern their subnets. They know the minimum amount of collateral requirements from their parent, and they need to ascribe to it, but from there on, is up to the developer of the subnet actor to determine the state of the subnet. 

I personally don't think having the subnet actor in an `Inactive` state while the gateway is `Active` in the gateway is an issue. What this scenario is stating is that the subnet has the collateral requirements to interact with IPC, but for some reason, the policy implemented by the subnet developers in its subnet actor requires additional requirements (in terms of collateral, minimum number of validators, etc) to operate normally. Subnet actors should worry themselves that their requirements are a superset of the minimum ones set by the parent, or they could end up in weird places, but I think this situation should be accepted and actually enforced.

Even more, if you think of our design, what Fendermint and Lotus do for their normal operation is look at the state of the subnet. They only thing for which they look to the gateway (of the parent) is when they need to route cross-net primitives, but for this to happen, the chain should already have been bootstrapped and the chain needs to be making progress.

**Comment by adlrocha** [comment]
> That after calling join on a subnet, the validator is not added to the validator set, but their collateral is. Maybe I am not familiar enough with the specifics of join, but this seems counter intuitive.

This is because according to the minimum collateral requirements specified by the subnet, the user is adding to the stake to protect the subnet with additional collateral, but hasn't put enough collateral to earn validator rights (I agree that this may be overcomplicating a bit the model, and we should maybe consider not accepting users to join if they don't put the minimum collateral set for the subnet). This may improve things and prevent potential misconfigurations (if we agree on this one I will add an issue to track it). 

> I expected that the sum total would have to clear this threshold

This should be the case, and if this is not it is definitely a bug somewhere.
